src/macro-context.js

```diff
@@ -172,18 +172,29 @@ export function unwrap(x) {
   return x;
 }
 
+function cloneEnforester(enf) {
+  const { rest, prev, context } = enf;
+  return new Enforester(rest, prev, context);
+}
+
+function Marker () {}
+
 /*
 ctx :: {
   of: (Syntax) -> ctx
   next: (String) -> Syntax or Term
 }
 */
 export default class MacroContext {
+
   constructor(enf, name, context, useScope, introducedScope) {
+    const startMarker = new Marker();
+    const startEnf = cloneEnforester(enf);
     const priv = {
-      backup: enf,
       name,
-      context
+      context,
+      startMarker,
+      markers: new Map([[startMarker, enf]]),
```

Comment:
I didn't think `WeakMap` necessary as the `markers` doesn't outlive the `MacroContext` in which it's contained.